src/test/fuzz/block_index.cpp

```diff
@@ -0,0 +1,134 @@
+uint32_t g_bits{0x1d00ffff};
+
+bool operator==(const CBlockFileInfo& a, const CBlockFileInfo& b)
+{
+    return a.nBlocks == b.nBlocks &&
+        a.nSize == b.nSize &&
+        a.nUndoSize == b.nUndoSize &&
+        a.nHeightFirst == b.nHeightFirst &&
+        a.nHeightLast == b.nHeightLast &&
+        a.nTimeFirst == b.nTimeFirst &&
+        a.nTimeLast == b.nTimeLast;
+}
+
+CBlockHeader ConsumeBlockHeader(FuzzedDataProvider& provider)
+{
+    CBlockHeader header;
+    header.nVersion = provider.ConsumeIntegral<decltype(header.nVersion)>();
+    header.hashPrevBlock = g_block_hash;
+    header.hashMerkleRoot = g_block_hash;
+    header.nTime = provider.ConsumeIntegral<decltype(header.nTime)>();
+    header.nBits = g_bits;
+    header.nNonce = provider.ConsumeIntegral<decltype(header.nNonce)>();
+    return header;
+}
+
+} // namespace
+
+void init_block_index()
+{
+    static const auto testing_setup = MakeNoLogFileContext<>(ChainType::MAIN);
+    g_setup = testing_setup.get();
+}
+
+FUZZ_TARGET(block_index, .init = init_block_index)
+{
+    FuzzedDataProvider fuzzed_data_provider{buffer.data(), buffer.size()};
+    auto block_index = kernel::BlockTreeDB(DBParams{
+        .path = "", // Memory only.
+        .cache_bytes = 1 << 20, // 1MB.
+        .memory_only = true,
+    });
+
+    // Generate a number of block files to be stored in the index.
+    int files_count = fuzzed_data_provider.ConsumeIntegralInRange(1, 100);
+    std::vector<std::unique_ptr<CBlockFileInfo>> files;
+    files.reserve(files_count);
+    std::vector<std::pair<int, const CBlockFileInfo*>> files_info;
+    files_info.reserve(files_count);
+    for (int i = 0; i < files_count; i++) {
+        if (auto file_info = ConsumeDeserializable<CBlockFileInfo>(fuzzed_data_provider)) {
+            files.push_back(std::make_unique<CBlockFileInfo>(std::move(*file_info)));
+            files_info.emplace_back(i, files.back().get());
+        } else {
+            return;
+        }
+    }
+
+    // Generate a number of block headers to be stored in the index.
+    int blocks_count = fuzzed_data_provider.ConsumeIntegralInRange(files_count * 10, files_count * 100);
+    std::vector<std::unique_ptr<CBlockIndex>> blocks;
+    blocks.reserve(blocks_count);
+    std::vector<const CBlockIndex*> blocks_info;
+    blocks_info.reserve(blocks_count);
+    for (int i = 0; i < blocks_count; i++) {
+        CBlockHeader header{ConsumeBlockHeader(fuzzed_data_provider)};
+        blocks.push_back(std::make_unique<CBlockIndex>(std::move(header)));
+        blocks.back()->phashBlock = &g_block_hash;
+        blocks_info.push_back(blocks.back().get());
+    }
+
+    // Store these files and blocks in the block index. It should not fail.
+    assert(block_index.WriteBatchSync(files_info, files_count - 1, blocks_info));
+
+    // We should be able to read every block file info we stored. Its value should correspond to
+    // what we stored above.
+    CBlockFileInfo info;
```

Comment:
nit (feel free to ignore):

```diff
diff --git a/src/test/fuzz/block_index.cpp b/src/test/fuzz/block_index.cpp
index 7885eda747..bc073e4930 100644
--- a/src/test/fuzz/block_index.cpp
+++ b/src/test/fuzz/block_index.cpp
@@ -93,8 +93,8 @@ FUZZ_TARGET(block_index, .init = init_block_index)
 
     // We should be able to read every block file info we stored. Its value should correspond to
     // what we stored above.
-    CBlockFileInfo info;
     for (const auto& [n, file_info]: files_info) {
+        CBlockFileInfo info;
         assert(block_index.ReadBlockFileInfo(n, info));
         assert(info == *file_info);
     }
```

Reply:
Why?

Reply:
Nevermind, my bad. It's better to reuse the same variable.